Adding actuation model with TN curve

CarlottaSartore opened the pull request:
<!-- readthedocs-preview jaxsim start -->
----
📚 Documentation preview 📚: https://jaxsim--405.org.readthedocs.build//405/

<!-- readthedocs-preview jaxsim end -->

younik reviewed (approved):
great!

flferretti commented on src/jaxsim/rbda/actuation/common.py:
```diff
@@ -0,0 +1,11 @@
+
+class ActuationParams(JaxsimDataclass):
+    """
+    Abstract class representing the parameters of an actuation model.
```
This is not an abstract class
```suggestion
    Parameters class for the actuation model.
```

flferretti commented on src/jaxsim/api/model.py:
```diff
@@ -272,6 +284,7 @@ def build(
             terrain=terrain,
             contact_model=contact_model,
             contact_params=contact_params,
+            actuation_params=actuation_params,
```
This attribute should be passed also in `js.model.reduce` otherwise it won't be inherited

CarlottaSartore replied:
Thanks, nice catch  ! Done in 49aa8db48affc36975abc0e2673898b41b05818e

flferretti reviewed (commented):
Thanks @CarlottaSartore! I added some comments. By the way, I think that we could just use `api.actuation_model` module instead of creating a new one which may cause confusion

flferretti reviewed (changes requested):
By the way, I think that we could just use `api.actuation_model` module instead of creating a new one which may cause confusion

CarlottaSartore commented:
About 

> By the way, I think that we could just use api.actuation_model module instead of creating a new one which may cause confusion

I have decided to create a new module because I thought of possible extensions in a near future with more complex actuation models, so I decided to replicate the same structure as for contact, which has an api.contact but also a module contacts where we implement the different contact models. I imagine the same for the actuation. 
What do you think @flferretti ?  Also @xela-95  with whom we created the first actuation class

CarlottaSartore commented:
> Thanks!
> 
> In future we could add simple unit tests in which we set a reference torque and check that this get limited by the actuation model.

Yes ! This could have sense ! I was struggling to identify a unit test to test this feature,  but what you are proposing could do the job !

CarlottaSartore commented:
> Thanks!
> 
> In future we could add simple unit tests in which we set a reference torque and check that this get limited by the actuation model.
> 
> Yes ! This could have sense ! I was struggling to identify a unit test to test this feature, but what you are proposing could do the job !

In the end I add the CI test in this PR  ! Feel free to review it , While doing it I noticed that: 

- We do not store in data the joint torque that we actually apply to the dynamics, it could have sense to store it   so that we can check the difference in between the desiredata and the actual torque applied. 
- I added a test for the actuation; we could also move the test for the joint limits since it is linked to the joint.  I did not do this in this PR since it would mean moving a few things around out of the scope of the PR,I added this comment only for for future reference

xela-95 commented:
> In the end I add the CI test in this PR ! Feel free to review it , While doing it I noticed that:
> 
> * We do not store in data the joint torque that we actually apply to the dynamics, it could have sense to store it   so that we can check the difference in between the desiredata and the actual torque applied.
> * I added a test for the actuation; we could also move the test for the joint limits since it is linked to the joint.  I did not do this in this PR since it would mean moving a few things around out of the scope of the PR,I added this comment only for for future reference

Super!! Simple and effective as it should be! and yes, unit tests sometimes help to discover things we miss :)

flferretti commented:
> About
> 
> > By the way, I think that we could just use api.actuation_model module instead of creating a new one which may cause confusion
> 
> I have decided to create a new module because I thought of possible extensions in a near future with more complex actuation models, so I decided to replicate the same structure as for contact, which has an api.contact but also a module contacts where we implement the different contact models. I imagine the same for the actuation. What do you think @flferretti ? Also @xela-95 with whom we created the first actuation class

Alright! I was thinking that since the class is not in the `api`, it might be counter intuitive for the user to use that, but we can modify this behavior in the future since we have the same issue with the contact models. Thank you!

flferretti commented on src/jaxsim/rbda/actuation/common.py:
```diff
@@ -0,0 +1,12 @@
+    Parameters class for the actuation model.
+    """
+
+    τ_max: jtp.Float = 3000.0  # Max torque (Nm)
+    ω_th: jtp.Float = 30.0  # Threshold speed (rad/s)
+    ω_max: jtp.Float = 100.0  # Max speed for torque drop-off (rad/s)
```
I'm usually a fan of unicode chars, but I'm super convinced of using them with user-exposed variables, e.g. in IPython these attribute would be difficult to access/create. Can we switch to ascii chars?

```suggestion
    max_torque: jtp.Float = 3000.0  # (Nm)
    threshold_speed: jtp.Float = 30.0  # (rad/s)
    dropoff_speed: jtp.Float = 100.0  # (rad/s)
```

CarlottaSartore replied:
Done in ba556cd4e97d43b10710ad06c9b3af842ddbcd8b

flferretti commented on src/jaxsim/rbda/actuation/common.py:
```diff
@@ -0,0 +1,12 @@
+    Parameters class for the actuation model.
+    """
+
+    τ_max: jtp.Float = 3000.0  # Max torque (Nm)
+    ω_th: jtp.Float = 30.0  # Threshold speed (rad/s)
+    ω_max: jtp.Float = 100.0  # Max speed for torque drop-off (rad/s)
```
Can we use `dataclasses.field` as these attributes are type-annotated and so that they appear in `__repr__()`?
```suggestion
    τ_max: jtp.Float = dataclasses.field(default=3000.0)  # Max torque (Nm)
    ω_th: jtp.Float = dataclasses.field(default=30.0)  # Threshold speed (rad/s)
    ω_max: jtp.Float = dataclasses.field(
        default=100.0
    )  # Max speed for torque drop-off (rad/s)
```

CarlottaSartore replied:
done in ba556cd4e97d43b10710ad06c9b3af842ddbcd8b

CarlottaSartore replied:
This actually  does not make the field instantiate correctly  see 
     
     ```
     """
        Compute the torque limits using the tn curve.
    
        Args:
            model: The model to consider.
            data: The data of the considered model.
    
        Returns:
            The torque limits.
        """
    
        τ_max = model.actuation_params.torque_max  # Max torque (Nm)
        ω_th = model.actuation_params.omega_th  # Threshold speed (rad/s)
        ω_max = model.actuation_params.omega_max  # Max speed for torque drop-off (rad/s)
        abs_vel = jnp.abs(data.joint_velocities)
        τ_lim = jnp.where(
>           abs_vel <= ω_th,
            τ_max,
            jnp.where(
                abs_vel <= ω_max, τ_max * (1 - (abs_vel - ω_th) / (ω_max - ω_th)), 0.0
            ),
        )
E       TypeError: '<=' not supported between instances of 'DynamicJaxprTracer' and 'Field'
E       --------------------
E       For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.

```

Do you know why this happens with fields ?

flferretti replied:
Sorry, I haven't seen this before writing https://github.com/ami-iit/jaxsim/pull/405#issuecomment-2797047047

CarlottaSartore replied:
Should be done now !

flferretti reviewed (commented):
Last few comments, sorry

flferretti commented:
> @flferretti I should have implemented all the comments! Let me know if I can proceed with the merge !

Thanks a lot @CarlottaSartore! The error in CI can be fixed decorating the class with `@jax_dataclasses.pytree_dataclass`

flferretti commented:
@CarlottaSartore it looks like there is a slowdown of the `test_benchmark.py::test_relaxed_rigid_contact_model[128]`, can we wait a bit until we understand the reason? I don't think it's related to your PR though

CarlottaSartore commented:
> @CarlottaSartore it looks like there is a slowdown of the `test_benchmark.py::test_relaxed_rigid_contact_model[128]`, can we wait a bit until we understand the reason? I don't think it's related to your PR though

Yes sure ! I see that  the same happens here https://github.com/ami-iit/jaxsim/pull/407#pullrequestreview-2760079629. I will comment there also !

flferretti commented:
> > @CarlottaSartore it looks like there is a slowdown of the test_benchmark.py::test_relaxed_rigid_contact_model[128], can we wait a bit until we understand the reason? I don't think it's related to your PR though
> 
> 
> 
> There is no slowdown anymore, as can be seen here https://github.com/ami-iit/jaxsim/pull/405#pullrequestreview-2760098248, the slow down was most likely due to some CI problem rather than the PR itself. I think we can merge it now! What do you think @flferretti  ? 

I agree. Good to go for me, thanks!

CarlottaSartore commented:
Merging
